File: examples_cpp/src/generators/gcopter_generator.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file gcopter_generator.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "generators/gcopter_generator.hpp"

#include <yaml-cpp/yaml.h>

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <utility>
#include <vector>

#include "utils/example_config_utils.hpp"
#include "utils/utils.hpp"

namespace mpc_examples::adapters {

namespace {

constexpr double kMinHorizontalSpeedForYaw = 0.5;  // [m/s]
constexpr double kMaxYawRateRefRadPerSec   = 1.0;  // [rad/s]

double wrapToPi(double x) {
  while (x > M_PI) {
    x -= 2.0 * M_PI;
  }
  while (x < -M_PI) {
    x += 2.0 * M_PI;
  }
  return x;
}

}  // namespace

GcopterGenerator::GcopterGenerator(const Config& cfg) : cfg_(cfg) {
  if (cfg_.corridor_margin <= 0.0) {
    throw std::invalid_argument("GcopterGenerator: corridor_margin must be > 0.");
  }
}

GcopterGenerator::Config GcopterGenerator::loadConfigFromYaml(const std::string& path) {
  const YAML::Node root = detail::loadYamlRoot(path);
  Config cfg;

  const auto req = [](const YAML::Node& n, const char* key) {
    return detail::readDoubleRequired(n[key], key);
  };

  if (!root["trajectory_generator"] || !root["drone_params"] || !root["drone_limits"] ||
      !root["optimization"]) {
    throw std::runtime_error(
        "GcopterGenerator: YAML must contain trajectory_generator, drone_params, "
        "drone_limits and optimization sections.");
  }

  const YAML::Node tg  = root["trajectory_generator"];
  const YAML::Node dp  = root["drone_params"];
  const YAML::Node dl  = root["drone_limits"];
  const YAML::Node opt = root["optimization"];

  cfg.corridor_margin = req(tg, "corridor_margin");

  cfg.drone_params.mass                = req(dp, "mass");
  cfg.drone_params.gravity             = req(dp, "gravity");
  cfg.drone_params.horizontal_drag     = req(dp, "horizontal_drag");
  cfg.drone_params.vertical_drag       = req(dp, "vertical_drag");
  cfg.drone_params.parasitic_drag      = req(dp, "parasitic_drag");
  cfg.drone_params.speed_smooth_factor = req(dp, "speed_smooth_factor");

  // drone_limits.max_velocity is filled in initialize() from ExampleConfig::max_speed.
  cfg.drone_limits.max_body_rate  = req(dl, "max_body_rate");
  cfg.drone_limits.max_tilt_angle = req(dl, "max_tilt_angle");
  cfg.drone_limits.min_thrust     = req(dl, "min_thrust");
  cfg.drone_limits.max_thrust     = req(dl, "max_thrust");

  cfg.optimization.time_weight      = req(opt, "time_weight");
  cfg.optimization.position_weight  = req(opt, "position_weight");
  cfg.optimization.velocity_weight  = req(opt, "velocity_weight");
  cfg.optimization.body_rate_weight = req(opt, "body_rate_weight");
  cfg.optimization.tilt_weight      = req(opt, "tilt_weight");
  cfg.optimization.thrust_weight    = req(opt, "thrust_weight");
  if (opt["smoothing_eps"]) {
    cfg.optimization.smoothing_eps =
        detail::readDoubleRequired(opt["smoothing_eps"], "smoothing_eps");
  }
  if (opt["integral_resolution"]) {
    cfg.optimization.integral_resolution = opt["integral_resolution"].as<int>();
  }
  if (opt["rel_cost_tol"]) {
    cfg.optimization.rel_cost_tol = detail::readDoubleRequired(opt["rel_cost_tol"], "rel_cost_tol");
  }
  return cfg;
}

void GcopterGenerator::initialize(const mav_model::State& initial_state,
                                  const ExampleConfig& example_cfg) {
  if (example_cfg.max_speed <= 0.0) {
    throw std::invalid_argument(
        "GcopterGenerator: ExampleConfig::max_speed must be > 0 "
        "(set in config_example.yaml).");
  }
  path_facing_  = example_cfg.path_facing;
  has_prev_t_   = false;
  prev_t_       = 0.0;
  yaw_ref_hold_ = quaternionToEuler(initial_state.getOrientationVector()).z();

  cfg_.drone_limits.max_velocity = example_cfg.max_speed;

  hold_pos_        = initial_state.getPositionVector();
  target_wp_       = hold_pos_;
  duration_        = 0.0;
  t_segment_start_ = 0.0;
  has_plan_        = false;

  // Construct the solver once; it is reused across segments via generate().
  gcopter_lib::GeneratorConfig generator_cfg;
  generator_cfg.params                       = cfg_.drone_params;
  generator_cfg.limits                       = cfg_.drone_limits;
  generator_cfg.optimization                 = cfg_.optimization;
  generator_cfg.optimization.corridor_margin = cfg_.corridor_margin;
  ctrl_ = std::make_unique<gcopter_lib::TrajectoryGenerator>(generator_cfg);
}

void GcopterGenerator::onWaypointChanged(const Eigen::Vector3d& next_waypoint,
                                         const mav_model::State& state,
                                         double t_start) {
  const Eigen::Vector3d p0 = state.getPositionVector();
  target_wp_               = next_waypoint;
  hold_pos_                = next_waypoint;
  t_segment_start_         = t_start;

  if ((next_waypoint - p0).norm() < 1e-6) {
    has_plan_ = false;
    duration_ = 0.0;
    return;
  }

  // Two-waypoint hop. MINCO + L-BFGS converges fine for any non-degenerate
  // segment; the pure-vertical degeneracy is handled inside gcopter_lib via
  // OptimizationConfig::vertical_perturbation, so the adapter just hands
  // (start, end) to the solver and trusts it to converge.
  std::vector<gcopter_lib::Waypoint> wps(2);
  wps[0].position = p0;
  wps[1].position = next_waypoint;
  has_plan_       = ctrl_->generate(wps, cfg_.drone_limits.max_velocity);
  if (!has_plan_) {
    // Fall back to a static setpoint at next_waypoint. The comparison remains
    // meaningful: gcopter failing here is an intrinsic property of batch
    // optimisation on a degenerate 2-point segment.
    duration_ = 0.0;
    return;
  }
  duration_ = ctrl_->duration();
}

void GcopterGenerator::update(double t, const mav_model::State& /*state*/) {
  const double dt = has_prev_t_ ? std::max(t - prev_t_, 0.0) : 0.0;
  prev_t_         = t;
  has_prev_t_     = true;

  if (!has_plan_) {
    return;
  }
  const double t_local = std::clamp(t - t_segment_start_, 0.0, duration_);

  if (path_facing_) {
    const Eigen::Vector3d vel = ctrl_->velocity(t_local);
    const double horiz_speed  = vel.head<2>().norm();
    if (horiz_speed > kMinHorizontalSpeedForYaw) {
      const double yaw_target = std::atan2(vel.y(), vel.x());
      double yaw_delta        = wrapToPi(yaw_target - yaw_ref_hold_);
      const double step       = kMaxYawRateRefRadPerSec * dt;
      yaw_delta               = std::clamp(yaw_delta, -step, step);
      yaw_ref_hold_           = yaw_ref_hold_ + yaw_delta;
    }
  } else {
    yaw_ref_hold_ = 0.0;
  }
}

framework::ReferenceSample GcopterGenerator::evaluate(double t) const {
  framework::ReferenceSample sample;
  sample.yaw = yaw_ref_hold_;

  if (!has_plan_) {
    sample.position = hold_pos_;
    return sample;
  }
  const double t_rel  = t - t_segment_start_;
  const double t_eval = std::clamp(t_rel, 0.0, duration_);
  sample.position     = ctrl_->position(t_eval);
  sample.velocity     = ctrl_->velocity(t_eval);
  sample.acceleration = ctrl_->acceleration(t_eval);

  // Past the segment end, freeze on the endpoint with zero motion.
  if (t_rel >= duration_) {
    sample.velocity.setZero();
    sample.acceleration.setZero();
  }
  return sample;
}

}  // namespace mpc_examples::adapters

File: examples_py/examples_py/generators/gcopter_generator.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""GCOPTER trajectory adapter (point-to-point polynomial planner).

Python mirror of
``examples/adapters/trajectory_generators/gcopter/src/gcopter_generator.cpp``.

Point-to-point contract: the scheduler calls :meth:`on_waypoint_changed` with
the next target; at each transition, the L-BFGS solver is re-run with a
three-waypoint path ``[current_pos, midpoint, next_waypoint]`` to avoid the
L-BFGS "negative line-search step" failure mode on a single segment.
``evaluate`` samples the resulting polynomial using a segment-local time
``t - t_segment_start``.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import math
import os
from dataclasses import dataclass, field
from typing import Optional

import numpy as np
import yaml
from gcopterpy import GeneratorConfig
from gcopterpy.trajectory import (
    DroneLimits,
    DroneParameters,
    OptimizationConfig,
    TrajectoryGenerator,
    Waypoint,
)
from mavpy.model import State

from examples_py.framework import (
    ExampleConfig,
    ITrajectoryGenerator,
    ReferenceField,
    ReferenceSample,
    make_mask,
)


_MIN_HORIZONTAL_SPEED_FOR_YAW = 0.5   # [m/s]
_MAX_YAW_RATE_REF_RAD_PER_SEC = 1.0   # [rad/s]


@dataclass
class _DroneParamsSpec:
    mass: float = 1.0
    gravity: float = 9.81
    horizontal_drag: float = 0.0
    vertical_drag: float = 0.0
    parasitic_drag: float = 0.0
    speed_smooth_factor: float = 0.0


@dataclass
class _DroneLimitsSpec:
    max_velocity: float = 3.0
    max_body_rate: float = 3.0
    max_tilt_angle: float = 1.0
    min_thrust: float = 0.0
    max_thrust: float = 30.0


@dataclass
class _OptimizationSpec:
    time_weight: float = 1.0
    position_weight: float = 100.0
    velocity_weight: float = 10.0
    body_rate_weight: float = 1.0
    tilt_weight: float = 1.0
    thrust_weight: float = 1.0
    smoothing_eps: float = 1.0e-2
    integral_resolution: int = 8
    rel_cost_tol: float = 1.0e-3


@dataclass
class GcopterConfig:
    corridor_margin: float = 2.0
    drone_params: _DroneParamsSpec = field(default_factory=_DroneParamsSpec)
    drone_limits: _DroneLimitsSpec = field(default_factory=_DroneLimitsSpec)
    optimization: _OptimizationSpec = field(default_factory=_OptimizationSpec)


def _quat_to_yaw(q: np.ndarray) -> float:
    w, x, y, z = float(q[0]), float(q[1]), float(q[2]), float(q[3])
    siny_cosp = 2.0 * (w * z + x * y)
    cosy_cosp = 1.0 - 2.0 * (y * y + z * z)
    return math.atan2(siny_cosp, cosy_cosp)


def _wrap_to_pi(x: float) -> float:
    while x > math.pi:
        x -= 2.0 * math.pi
    while x < -math.pi:
        x += 2.0 * math.pi
    return x


def _require(node: dict, key: str, section: str) -> float:
    if key not in node:
        raise ValueError(f'gcopter config: {section}.{key} is required.')
    return float(node[key])


class GcopterGenerator(ITrajectoryGenerator):
    """Point-to-point GCOPTER adapter. Re-plans on every waypoint change."""

    def __init__(self, cfg: GcopterConfig) -> None:
        if cfg.corridor_margin <= 0.0:
            raise ValueError('GcopterGenerator: corridor_margin must be > 0.')
        self._cfg = cfg
        self._ctrl: Optional[TrajectoryGenerator] = None

        self._target_wp = np.zeros(3, dtype=float)
        self._hold_pos = np.zeros(3, dtype=float)
        self._duration = 0.0
        self._t_segment_start = 0.0
        self._has_plan = False
        self._path_facing = False

        self._yaw_ref_hold = 0.0
        self._prev_t = 0.0
        self._has_prev_t = False
        self._name = 'GcopterGenerator'

    @staticmethod
    def load_config_from_yaml(path: str) -> GcopterConfig:
        if not os.path.isfile(path):
            raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
        with open(path, 'r') as f:
            root = yaml.safe_load(f)
        if not isinstance(root, dict):
            raise ValueError('gcopter config: root must be a mapping.')
        for key in ('trajectory_generator', 'drone_params', 'drone_limits', 'optimization'):
            if key not in root or not isinstance(root[key], dict):
                raise ValueError(
                    f"gcopter config: '{key}' section must be present and a mapping.")

        tg = root['trajectory_generator']
        dp = root['drone_params']
        dl = root['drone_limits']
        opt = root['optimization']

        cfg = GcopterConfig()
        cfg.corridor_margin = _require(tg, 'corridor_margin', 'trajectory_generator')

        cfg.drone_params.mass = _require(dp, 'mass', 'drone_params')
        cfg.drone_params.gravity = _require(dp, 'gravity', 'drone_params')
        cfg.drone_params.horizontal_drag = _require(
            dp, 'horizontal_drag', 'drone_params')
        cfg.drone_params.vertical_drag = _require(
            dp, 'vertical_drag', 'drone_params')
        cfg.drone_params.parasitic_drag = _require(
            dp, 'parasitic_drag', 'drone_params')
        cfg.drone_params.speed_smooth_factor = _require(
            dp, 'speed_smooth_factor', 'drone_params')

        cfg.drone_limits.max_body_rate = _require(dl, 'max_body_rate', 'drone_limits')
        cfg.drone_limits.max_tilt_angle = _require(dl, 'max_tilt_angle', 'drone_limits')
        cfg.drone_limits.min_thrust = _require(dl, 'min_thrust', 'drone_limits')
        cfg.drone_limits.max_thrust = _require(dl, 'max_thrust', 'drone_limits')

        cfg.optimization.time_weight = _require(opt, 'time_weight', 'optimization')
        cfg.optimization.position_weight = _require(
            opt, 'position_weight', 'optimization')
        cfg.optimization.velocity_weight = _require(
            opt, 'velocity_weight', 'optimization')
        cfg.optimization.body_rate_weight = _require(
            opt, 'body_rate_weight', 'optimization')
        cfg.optimization.tilt_weight = _require(opt, 'tilt_weight', 'optimization')
        cfg.optimization.thrust_weight = _require(opt, 'thrust_weight', 'optimization')
        if 'smoothing_eps' in opt:
            cfg.optimization.smoothing_eps = float(opt['smoothing_eps'])
        if 'integral_resolution' in opt:
            cfg.optimization.integral_resolution = int(opt['integral_resolution'])
        if 'rel_cost_tol' in opt:
            cfg.optimization.rel_cost_tol = float(opt['rel_cost_tol'])
        return cfg

    def _build_native_drone_params(self) -> DroneParameters:
        dp = DroneParameters()
        src = self._cfg.drone_params
        dp.mass = src.mass
        dp.gravity = src.gravity
        dp.horizontal_drag = src.horizontal_drag
        dp.vertical_drag = src.vertical_drag
        dp.parasitic_drag = src.parasitic_drag
        dp.speed_smooth_factor = src.speed_smooth_factor
        return dp

    def _build_native_drone_limits(self) -> DroneLimits:
        dl = DroneLimits()
        src = self._cfg.drone_limits
        dl.max_velocity = src.max_velocity
        dl.max_body_rate = src.max_body_rate
        dl.max_tilt_angle = src.max_tilt_angle
        dl.min_thrust = src.min_thrust
        dl.max_thrust = src.max_thrust
        return dl

    def _build_native_optimization(self) -> OptimizationConfig:
        oc = OptimizationConfig()
        src = self._cfg.optimization
        oc.time_weight = src.time_weight
        oc.position_weight = src.position_weight
        oc.velocity_weight = src.velocity_weight
        oc.body_rate_weight = src.body_rate_weight
        oc.tilt_weight = src.tilt_weight
        oc.thrust_weight = src.thrust_weight
        oc.smoothing_eps = src.smoothing_eps
        oc.integral_resolution = src.integral_resolution
        oc.rel_cost_tol = src.rel_cost_tol
        oc.corridor_margin = self._cfg.corridor_margin
        return oc

    def _build_native_generator_config(self) -> GeneratorConfig:
        cfg = GeneratorConfig()
        cfg.params = self._build_native_drone_params()
        cfg.limits = self._build_native_drone_limits()
        cfg.optimization = self._build_native_optimization()
        return cfg

    def initialize(self, initial_state: State, example_cfg: ExampleConfig) -> None:
        if example_cfg.max_speed <= 0.0:
            raise ValueError(
                'GcopterGenerator: ExampleConfig.max_speed must be > 0 '
                '(set in config_example.yaml).')
        self._path_facing = bool(example_cfg.path_facing)
        self._has_prev_t = False
        self._prev_t = 0.0
        self._yaw_ref_hold = _quat_to_yaw(
            np.asarray(initial_state.orientation, dtype=float))

        self._cfg.drone_limits.max_velocity = float(example_cfg.max_speed)

        self._hold_pos = np.asarray(initial_state.position, dtype=float).copy()
        self._target_wp = self._hold_pos.copy()
        self._duration = 0.0
        self._t_segment_start = 0.0
        self._has_plan = False

        # Construct the solver once; reused across segments via generate().
        self._ctrl = TrajectoryGenerator(self._build_native_generator_config())

    def on_waypoint_changed(
        self, next_waypoint: np.ndarray, state: State, t_start: float,
    ) -> None:
        if self._ctrl is None:
            raise RuntimeError(
                'GcopterGenerator: initialize() must be called before on_waypoint_changed().')

        p0 = np.asarray(state.position, dtype=float)
        self._target_wp = np.asarray(next_waypoint, dtype=float).copy()
        self._hold_pos = self._target_wp.copy()
        self._t_segment_start = t_start

        if float(np.linalg.norm(self._target_wp - p0)) < 1e-6:
            self._has_plan = False
            self._duration = 0.0
            return

        # Two-waypoint hop. MINCO + L-BFGS converges fine for any non-degenerate
        # segment; the pure-vertical degeneracy is handled inside gcopter_lib
        # via OptimizationConfig::vertical_perturbation, so the adapter just
        # hands (start, end) to the solver and trusts it to converge.
        wps = [Waypoint(p0), Waypoint(self._target_wp)]
        ok = self._ctrl.generate(wps, self._cfg.drone_limits.max_velocity)
        if not ok:
            # Fall back to a static setpoint at next_waypoint.
            self._has_plan = False
            self._duration = 0.0
            return
        self._has_plan = True
        self._duration = float(self._ctrl.duration())

    def update(self, t: float, state: State) -> None:
        del state
        dt = max(t - self._prev_t, 0.0) if self._has_prev_t else 0.0
        self._prev_t = t
        self._has_prev_t = True

        if not self._has_plan or self._ctrl is None:
            return
        t_local = min(max(t - self._t_segment_start, 0.0), self._duration)

        if self._path_facing:
            vel = np.asarray(self._ctrl.velocity(t_local), dtype=float)
            horiz_speed = float(np.linalg.norm(vel[:2]))
            if horiz_speed > _MIN_HORIZONTAL_SPEED_FOR_YAW:
                yaw_target = math.atan2(float(vel[1]), float(vel[0]))
                yaw_delta = _wrap_to_pi(yaw_target - self._yaw_ref_hold)
                step = _MAX_YAW_RATE_REF_RAD_PER_SEC * dt
                yaw_delta = max(-step, min(step, yaw_delta))
                self._yaw_ref_hold += yaw_delta
        else:
            self._yaw_ref_hold = 0.0

    def evaluate(self, t: float) -> ReferenceSample:
        sample = ReferenceSample()
        sample.yaw = self._yaw_ref_hold

        if not self._has_plan or self._ctrl is None:
            sample.position = self._hold_pos.copy()
            return sample

        t_rel = t - self._t_segment_start
        t_eval = min(max(t_rel, 0.0), self._duration)
        sample.position = np.asarray(
            self._ctrl.position(t_eval), dtype=float).copy()
        sample.velocity = np.asarray(
            self._ctrl.velocity(t_eval), dtype=float).copy()
        sample.acceleration = np.asarray(
            self._ctrl.acceleration(t_eval), dtype=float).copy()

        if t_rel >= self._duration:
            sample.velocity.fill(0.0)
            sample.acceleration.fill(0.0)
        return sample

    def provided_reference_fields(self) -> ReferenceField:
        return make_mask([
            ReferenceField.POSITION,
            ReferenceField.VELOCITY,
            ReferenceField.ACCELERATION,
        ])

    def name(self) -> str:
        return self._name
